_posts/2019-07-30-microfrontends-holidaycheck-part2.md

```diff
@@ -0,0 +1,168 @@
+```html
+<script type="module" src="bundle.es6.js" async defer></script>
+<script noModule src="bundle.es5.js" async defer></script>
+```
+
+This means you have to build your application twice, once with an ES5 and once with an ES6 target.
+
+# CSS
+Keeping an eye on how much CSS is delivered is also important. The following things were relevant for us.
+
+* Remove unused CSS
+* Inline "critical" CSS (CSS which is necessary for the first impression)
+* Load the rest asynchronously and not render blocking
+
+To identify unused and therefore obsolete CSS we use the Coverage report from Chrome Dev Tools, which is 
+available in the *Source* tab. You can even click around on your page and the coverage updates in real time.
+
+<figure>
+    <figcaption>Inspecting code coverage of CSS & JS in Chrome Dev Tools</figcaption>
+    <img src="/img/posts/2019-07-30-microfrontends-holidaycheck-part2/css-coverage.png" alt="css-coverage" />
+</figure>
+
+How to *inline the critical CSS* of your page depends a lot on your tool chain and the amount of automation you need
+for it. But once you are able to do this, you can load the rest of the CSS asynchronously without blocking the rendering.
+This is done with a `preload` link which we change to a `stylesheet` link once it is loaded, which triggers the browser
+to apply the stylesheets. 
+
+```html
+<link rel="preload" type="text/css" href="bundle.css" as="style"  onload="this.onload=null;this.rel='stylesheet'">
+<noscript><link rel="stylesheet" type="text/css" href="bundle.css"></noscript>
+```
+
+Keep in mind that this only works for browsers that support preload links. For all others you need a polyfill,
+we use the one from [loadCSS](https://github.com/filamentgroup/loadCSS).
+
+## Server Side Rendering
+For a faster First Contentful Paint and First Meaningful Paint we decided to do SSR (Server Side Rendering).
+For most frameworks, for example React and Preact, it is quite simple to setup SSR, just check the docs 
+of the framework you use.
+
+When a doing SSR the browser can already render the page, even when the necessary Javascript is not loaded, parsed & executed yet.
+(also our SEO Team is very happy when we do SSR) Still SSR also has a drawback, by transferring the page prerendered 
+more content is sent to the Browser. So while SSR improves the paint metrics, it might have a negative impact on
+other speed metrics. 
+
+## Do not overdo
+In today's JS ecosystem it is quite a challenge to get speed right, however with performance budgets and some
+discipline to follow them it is possible to get a fast result. We often had to critically ask ourselves
+whether we prioritize developer experience or user experience (speed) higher. However I recommend to not 
+overdo it, delaying a feature for weeks just for the sake of some milli-seconds is often not sustainable.
+It is important to compromise at some point, but these compromises should be explicit decision and not
+something which happens accidentally. 
+
```

Comment:
Is it a Part 3 coming or you don't know yet? If yes, you could teaser it here as a wrapping up of the post.

Reply:
Actually I do not know yet. I will ask the readers for inspiration!